checker/src/org/checkerframework/checker/lock/qual/LockHeld.java

```diff
@@ -11,17 +16,26 @@
 import java.lang.annotation.Target;
 
 /**
- * {@link LockHeld} is a type annotation that indicates that an expression is
- * used as a lock and the lock is known to be held on the current thread.
+ * Indicates that an expression is used as a lock and the lock
+ * is known to be held on the current thread.
+ * <p>
+ *
+ * This annotation may not be written in source code; it is an
+ * implementation detail of the checker.
  *
  * @see LockPossiblyHeld
  * @checker_framework.manual #lock-checker Lock Checker
  */
 @TypeQualifier
 @SubtypeOf(LockPossiblyHeld.class) // This is the bottom type in this hierarchy
+@InvisibleQualifier
 @Documented
+// The following line is counterintuitive but it is needed to avoid errors of the kind:
+// found   : T[ extends @GuardedBy @LockPossiblyHeld Annotation super @GuardedByBottom @LockPossiblyHeld Void]
+// required: T[ extends @GuardedBy @LockPossiblyHeld Annotation super @GuardedByBottom @LockHeld Void]
+// whenever a variable of type T is assigned a null value.
+@ImplicitFor(trees = {Tree.Kind.NULL_LITERAL})
```

Comment:
Doesn't this produce a problem with flow-sensitive type refinement?

Object l = null;
refines the type of l to LockHeld. Can this cause an unsoundness? Or do you rely on a NullPointerException being raised in that case?
Instead of "a variable of type T" I would say "a type variable".